Implement `Channel.max_bias` and other small fixes

stavros11 opened the pull request:
Small enhancements to the QM driver that were implemented in #324 which is now closed as the runcard is no longer relevant. Particularly:

* Adds a `max_bias` attribute to the `Channel` object. When `bias` is set, the code checks if it is lower than the `max_bias` and raises an error otherwise. This is useful to limit the amount of power we send in the fluxlines, particularly when using amplifiers, as sending high power extensively may damage them. Note that `max_bias` should be set manually when defining the platform in order to use this feature.
* Implements value checks in QM sweepers: for example if ones tries to sweep frequencies beyond the instrument bandwidth (400MHz) an error is raised, because otherwise the instrument just returns wrong values.
* Pinned versions of `qm-qua` and `qualang-tools` to the currently latest versions, which work, to avoid issues like #340 in case a future release breaks something.

Checklist:
- [ ] Reviewers confirm new code works as expected.
- [ ] Tests are passing.
- [ ] Coverage does not decrease.
- [ ] Documentation is updated.

AleCandido commented on src/qibolab/instruments/qm.py:
```diff
@@ -709,7 +712,12 @@ def sweep_frequency(self, sweepers, qubits, qmsequence, relaxation_time):
+            f0 = int(pulse.frequency - lo_frequency)
+            freqs0.append(declare(int, value=f0))
+            # check if sweep is within the supported bandwidth [-400, 400] MHz
+            max_freq = max(abs(min(sweeper.values) + f0), abs(max(sweeper.values) + f0))
```
I'd expect:
```py
abs(min(sweeper.values) + f0) < abs(max(sweeper.values) + f0)
```
why this is not the case?

AleCandido replied:
Is it actually possible that:
```py
max(sweeper.values) + f0 < 0
```
?

stavros11 replied:
> I'd expect:
> 
> ```python
> abs(min(sweeper.values) + f0) < abs(max(sweeper.values) + f0)
> ```
> 
> why this is not the case?

`f0` is an IF so it can be negative, so for example if:
* `f0 = -200`
* `min(sweeper.values) = -100`
* `max(sweeper.values) = 100`

the above inequality is violated.

AleCandido commented on src/qibolab/instruments/qm.py:
```diff
@@ -752,22 +758,28 @@ def sweep_relative_phase(self, sweepers, qubits, qmsequence, relaxation_time):
+        for q in sweeper.qubits:
+            b0 = qubits[q].flux.bias
+            max_bias = qubits[q].flux.max_bias
+            max_value = max(abs(min(sweeper.values) + b0), abs(max(sweeper.values) + b0))
```
Same doubts as above.

stavros11 replied:
Same as above, `b0` can be negative.

AleCandido commented on src/qibolab/instruments/qm.py:
```diff
@@ -752,22 +758,28 @@ def sweep_relative_phase(self, sweepers, qubits, qmsequence, relaxation_time):
+            b0 = qubits[q].flux.bias
+            max_bias = qubits[q].flux.max_bias
+            max_value = max(abs(min(sweeper.values) + b0), abs(max(sweeper.values) + b0))
+            if max_bias is not None and max_value > max_bias:
+                raise_error(
+                    ValueError,
+                    f"Cannot sweep bias up to {max_value} because the maximum supported value is {max_bias}.",
+                )
```
Moreover, consider making a custom error `MaximumBiasedExceedError` or something similar.

AleCandido reviewed (changes requested):
Mostly good, only a single critical error in `.bias` property

stavros11 commented on src/qibolab/designs/channels.py:
```diff
@@ -35,19 +35,26 @@ class Channel:
         if self._bias is None:
             # operate qubits at their sweetspot unless otherwise stated
-            return self.qubit.sweetspot
+            self.bias = self.qubit.sweetspot
```
Why is this an infinite recursion loop? I think it is just calling the setter, not the property, but it may not be good practice, I am not sure.

The reason I am doing this here, is because I want the 

https://github.com/qiboteam/qibolab/blob/0bec714998aa303e281d939eae26a70c336a7672/src/qibolab/designs/channels.py#L56

check to run before setting the bias as the sweetspot value. Otherwise I will have to repeat the check both here and in the setter.

AleCandido replied:
However, choose the solution you prefer, but I'd suggest to you to split the condition, i.e.
https://github.com/qiboteam/qibolab/blob/0bec714998aa303e281d939eae26a70c336a7672/src/qibolab/designs/channels.py#L56
and give it a name. I.e. make a function of it.

I know it might seem silly to make such a short function, but:
- giving it a name it will be more explicit
- in principle you can also include the error, not only the condition
  https://github.com/qiboteam/qibolab/blob/0bec714998aa303e281d939eae26a70c336a7672/src/qibolab/instruments/qm.py#L771-L775

So you will have:
```py
def check_max_bias(max_value, max_bias):
    if max_bias is not None and max_value > max_bias:
        raise_error(
            ValueError,
            f"Cannot sweep bias up to {max_value} because the maximum supported value is {max_bias}.",
        )
```

And in this way you don't even have to repeat the error, and will be even consistent in `channels.py` and `qm.py` :)

stavros11 commented on setup.py:
```diff
@@ -67,8 +67,8 @@ def get_version():
             "pyvisa-py==0.5.3",
-            "qm-qua>=1.0.1",
-            "qualang-tools>=0.13.1",
+            "qm-qua==1.1.1",
```
That's fine with me, however what happens if they release a `qm-qua 1.1.2` that breaks something in the driver? The idea of pinning to the exact version with `==` was to avoid such issues as it ensures that we are using the version that works.

Of course it is not optimal but the main issues are:
1. It appears that these libraries are a bit unstable and it is very likely that new releases contain breaking changes,
2. Driver coverage in CI is very low and unless we use the label tests which run on the QPUs (which are also not available sometimes), driver code will not be tested. This makes it hard to spot these breaking changes.

stavros11 commented:
Thanks for the review @AleCandido. I have implemented the functions for `check_max_bias` and calculating the maximum sweep value and replaced `int` with `math.floor`. 

The only unresolved comment is regarding the versions in setup.py. I left the `==` mainly for the reason I wrote above (future releases of qm-qua may not work out-of-the-box with the driver as their API is not stable) and we also follow the same approach of pinning to exact versions for other similar dependencies too (qblox-instruments, etc.).
